Add support to the Permissions API for dedicated workers

RupinMittal opened the pull request:
#### fd42e2c17767290e383a082d15593c4b74f79717
<pre>
Add support to the Permissions API for dedicated workers
<a href="https://bugs.webkit.org/show_bug.cgi?id=243805">https://bugs.webkit.org/show_bug.cgi?id=243805</a>

Reviewed by Chris Dumez and Sihui Liu.

Permissions::query() should be exposed for both window and workers. Currently,
it is only exposed for window. This patch is the first step in exposing it for
all of the workers.

* LayoutTests/http/tests/notifications/permission/worker-permission-query-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/worker-permission-query.html: Added.
* LayoutTests/http/tests/notifications/permission/worker-permission-query.js: Added.
(onmessage):
* LayoutTests/imported/w3c/web-platform-tests/permissions/all-permissions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/permission-query.https.any.worker-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/NavigatorPermissions.cpp:
(WebCore::NavigatorPermissions::permissions):
* Source/WebCore/Modules/permissions/NavigatorPermissions.h:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
(WebCore::PermissionStatus::stateChanged):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/PermissionStatus.idl:
* Source/WebCore/Modules/permissions/PermissionStatusIdentifier.h: Added.
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::create):
(WebCore::Permissions::Permissions):
(WebCore::Permissions::navigator):
(WebCore::Permissions::query):
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/Modules/permissions/Permissions.idl:
* Source/WebCore/Modules/permissions/WorkerNavigator+Permissions.idl: Copied from Source/WebCore/Modules/permissions/Permissions.idl.
* Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp: Copied from Source/WebCore/Modules/permissions/NavigatorPermissions.cpp.
(WebCore::WorkerNavigatorPermissions::WorkerNavigatorPermissions):
(WebCore::WorkerNavigatorPermissions::permissions):
(WebCore::WorkerNavigatorPermissions::from):
(WebCore::WorkerNavigatorPermissions::supplementName):
* Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.h: Copied from Source/WebCore/Modules/permissions/NavigatorPermissions.h.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::addObserver):
(WebKit::WebPermissionController::removeObserver):
(WebKit::WebPermissionController::permissionChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:

Canonical link: <a href="https://commits.webkit.org/253447@main">https://commits.webkit.org/253447@main</a>
</pre>

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any.worker-expected.txt:
```diff
@@ -27,21 +27,21 @@ FAIL PermissionStatus interface: existence and properties of interface prototype
-FAIL Permissions interface: calling query(object) on navigator.permissions with too few arguments must throw TypeError assert_equals: wrong typeof object expected "object" but got "undefined"
-FAIL WorkerNavigator interface: attribute permissions assert_true: The prototype object must have a property "permissions" expected true got false
-FAIL WorkerNavigator interface: navigator must inherit property "permissions" with the proper type assert_inherits: property "permissions" not found in prototype chain
+FAIL PermissionStatus must be primary interface of permissionStatus assert_own_property: self does not have own property "PermissionStatus" expected property "PermissionStatus" missing
```
Please fix the `Exposed=` in PermissionStatus.idl to fix this subtest.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/mediacapabilities/WorkerNavigatorMediaCapabilities.cpp:
```diff
@@ -61,6 +61,7 @@ MediaCapabilities& WorkerNavigatorMediaCapabilities::mediaCapabilities(WorkerNav
 
 MediaCapabilities& WorkerNavigatorMediaCapabilities::mediaCapabilities() const
 {
+    WTFReportBacktraceWithPrefix("[MediaCapabilities] ");
```
Should remove.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -25,7 +25,11 @@
 
 #pragma once
 
+#include "ClientOrigin.h"
```
Should probably use a forward declaration instead.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -25,7 +25,11 @@
 #pragma once
 
+#include "ClientOrigin.h"
+#include "PermissionDescriptor.h"
```
ditto.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -47,19 +57,47 @@ Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context,
+        allPermissionStatusObjects().add(m_identifier, this);
+    }
+
+    callOnMainThreadAndWait([this, &context]() { 
```
Why are we waiting here? This should very likely be a `ensureOnMainThread()` call.

Also, it is not OK to capture `this` and use it on another thread.

RupinMittal replied:
Fixed!

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -47,19 +57,47 @@ Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context,
+        allPermissionStatusObjects().remove(m_identifier);
+    }
+
+    callOnMainThreadAndWait([this]() { 
```
ensureOnMainThread(). I don't think we should be waiting at all.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -47,19 +57,47 @@ Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context,
+}
+
+void PermissionStatus::stateChanged(PermissionStatusIdentifier statusIdentifier, ScriptExecutionContextIdentifier contextIdentifier, PermissionState newState)
+{
```
Please add a `ASSERT(isMainThread());` at the beginning of this function.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -47,19 +57,47 @@ Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context,
+
+void PermissionStatus::stateChanged(PermissionStatusIdentifier statusIdentifier, ScriptExecutionContextIdentifier contextIdentifier, PermissionState newState)
+{
+    ScriptExecutionContext::ensureOnContextThread(contextIdentifier, [&statusIdentifier, newState](auto&) mutable {
```
You shouldn't capture `statusIdentifier` by reference when passing it to another thread. This isn't safe.
Pass by value instead.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -46,6 +48,10 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
 
+    WEBCORE_EXPORT static void stateChanged(PermissionStatusIdentifier, ScriptExecutionContextIdentifier, PermissionState);
+    // PermissionObserver
+    void stateChanged(PermissionState) final;
```
Does this really need to be public?

RupinMittal replied:
No, it doesn't. Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatusIdentifier.h:
```diff
@@ -0,0 +1,35 @@
+namespace WebCore {
+
+enum PermissionStatusIdentifierType { };
+using PermissionStatusIdentifier = ObjectIdentifier<PermissionStatusIdentifierType >;
```
nit: extra space before `<`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
 }
 
-void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DOMPromiseDeferred<IDLInterface<PermissionStatus>>&& promise)
+void Permissions::query(ScriptExecutionContext& context, JSC::Strong<JSC::JSObject> permissionDescriptorValue, DOMPromiseDeferred<IDLInterface<PermissionStatus>>&& promise)
```
Passing the `context` here seems wrong. This may not be the same context as Permissions's `context`. The one in parameter (coming from `CallWith` in IDL) is the caller's context. There spec says to use `this`'s relevant global object:
https://w3c.github.io/permissions/#dom-permissions-query

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
-        });
-        return;
+    if (is<Document>(context)) {
```
`if (document) {`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        });
-        return;
+    if (is<Document>(context)) {
+        if (!isAllowedByFeaturePolicy(downcast<Document>(context), parameterDescriptor.name)) {
```
Use `document` directly.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        return;
+    if (is<Document>(context)) {
+        if (!isAllowedByFeaturePolicy(downcast<Document>(context), parameterDescriptor.name)) {
+            context.postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
```
Why are we posting a task here? Why not resolve the promise right away? Is the spec saying to queue a task? Note that even if the spec was saying to queue a task, we'd use the HTML event loop, not postTask() for this.

RupinMittal replied:
You're right. We don't need to post a task here--this is older code, not sure what the logic was when it was written. Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
+    if (is<WorkerGlobalScope>(context)) {
```
Could use `if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope >(context)) {`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+    if (is<WorkerGlobalScope>(context)) {
+        auto* workerGlobalScope = downcast<WorkerGlobalScope>(&context);
+
+        auto completionHandler = [originData, parameterDescriptor, identifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
```
identifier is too vague, let's call this `contextIdentifier`.

cdumez replied:
`originData = WTFMove(originData).isolatedCopy()` here since you're passing it to another thread.

cdumez replied:
Please `ASSERT(isMainThread());` at the beginning of this lambda.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+        auto* workerGlobalScope = downcast<WorkerGlobalScope>(&context);
+
+        auto completionHandler = [originData, parameterDescriptor, identifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *downcast<Document>(context).page(), [&context, identifier, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
```
I don't think it is safe to assume that `downcast<Document>(context).page()` is non-null here.

cdumez replied:
You cannot capture context by reference since the lambda will run asynchronously. You'll have a Ref the context and capture the Ref<>.

cdumez replied:
WTFMove(originData)

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+        auto completionHandler = [originData, parameterDescriptor, identifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *downcast<Document>(context).page(), [&context, identifier, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                if (!context.globalObject())
+                    return;
```
You're failing to resolve the promise here before returning.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+                if (!context.globalObject())
+                    return;
+
+                context.postTask([identifier, parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto&) mutable {
```
Why are we doing this? I don't think it is needed..

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+                    return;
+
+                context.postTask([identifier, parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto&) mutable {
+                    ScriptExecutionContext::postTaskTo(identifier, [parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto& context) mutable {
```
No need to WTFMove(permissionState), just capture `permissionState`.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+                context.postTask([identifier, parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto&) mutable {
+                    ScriptExecutionContext::postTaskTo(identifier, [parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto& context) mutable {
+                        if (!permissionState)
+                            promise.reject(Exception { NotSupportedError });
```
It would be nice to provide a helpful error message as second parameter to the Exception constructor.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+        return;
+    }
+
+    PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [&context, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
```
WTFMove(originData)

cdumez replied:
You cannot capture an object by reference in a lambda that runs asynchronously. You'll have to ref the context.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+
+    PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [&context, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+        if (!context.globalObject())
             return;
```
You're failing to resolve the promise here.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -75,62 +80,74 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
             return;
 
-        context->postTask([parameterDescriptor, promise = WTFMove(promise),  permissionState = WTFMove(permissionState)](auto& context) mutable {
+        context.postTask([parameterDescriptor, promise = WTFMove(promise), permissionState = WTFMove(permissionState)](auto& context) mutable {
```
What's this. It shouldn't be needed. The spec doesn't say to queue a task to resolve the promise. It just says to resolve the promise. Even if it said to queue a task, we'd use the HTML event loop, not `postTask()`.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.h:
```diff
@@ -36,24 +36,24 @@ class JSObject;
-
-    WeakPtr<Navigator> m_navigator;
+    explicit Permissions(NavigatorBase&);
+    WeakPtr<NavigatorBase> m_navigator;
```
Please add a blank line before the data members, like there used to be.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.idl:
```diff
@@ -29,7 +29,7 @@
+    Exposed=(Window,DedicatedWorker)
 ] interface Permissions {
-    Promise<PermissionStatus> query(object permissionDesc);
+    [CallWith=CurrentScriptExecutionContext] Promise<PermissionStatus> query(object permissionDesc);
```
Likely wrong, see comment above.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigator+Permissions.idl:
```diff
@@ -0,0 +1,32 @@
+    EnabledBySetting=PermissionsAPIEnabled,
+    ImplementedBy=WorkerNavigatorPermissions,
+] partial interface WorkerNavigator {
+  [SameObject] readonly attribute Permissions permissions;
```
indentation seems wrong, missing 2 spaces.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp:
```diff
@@ -0,0 +1,73 @@
+WorkerNavigatorPermissions::WorkerNavigatorPermissions(WorkerNavigator& navigator)
+    : m_navigator(navigator)
+{
+    WTFLogAlways("%s", __PRETTY_FUNCTION__);
```
Should remove

youssefsoli replied:
If you want to still log, it might make sense to add an entry to `Logging.h` with `Permissions` and then you can do `LOG(Permissions, ...)`

RupinMittal replied:
Ah yeah, this is a mistake. I forgot to remove this debugging log before opening the pr. But thank for the suggestion about adding an entry to Logging.h, I'll keep that in mind! Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp:
```diff
@@ -0,0 +1,73 @@
+
+RefPtr<Permissions> WorkerNavigatorPermissions::permissions(WorkerNavigator& navigator)
+{
+    WTFLogAlways("%s", __PRETTY_FUNCTION__);
```
Should remove

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp:
```diff
@@ -0,0 +1,73 @@
+    return WorkerNavigatorPermissions::from(navigator)->permissions();
+}
+
+RefPtr<Permissions> WorkerNavigatorPermissions::permissions()
```
Should return a `Permissions&`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp:
```diff
@@ -0,0 +1,73 @@
+
+WorkerNavigatorPermissions* WorkerNavigatorPermissions::from(WorkerNavigator& navigator)
+{
+    WTFLogAlways("%s", __PRETTY_FUNCTION__);
```
Should remove.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp:
```diff
@@ -0,0 +1,73 @@
+
+const char* WorkerNavigatorPermissions::supplementName()
+{
+    WTFLogAlways("%s", __PRETTY_FUNCTION__);
```
Should remove.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/page/WorkerNavigator.cpp:
```diff
@@ -34,6 +34,9 @@ WorkerNavigator::WorkerNavigator(ScriptExecutionContext& context, const String&
     , m_userAgent(userAgent)
     , m_isOnline(isOnline)
 {
+    WTFLogAlways("%s", __PRETTY_FUNCTION__);
```
Should remove all 3 lines.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -121,21 +122,24 @@ void WebPermissionController::tryProcessingRequests()
 }
 
-void WebPermissionController::addObserver(WebCore::PermissionObserver& observer)
+void WebPermissionController::addObserver(WebCore::PermissionStatusIdentifier statusIdentifier, WebCore::ClientOrigin& origin, WebCore::PermissionDescriptor& descriptor, WebCore::ScriptExecutionContextIdentifier contextIdentifier)
```
Let's use a const reference for `origin` instead of a non-const one.

No need to pass `descriptor` by reference, it is just an enum type.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -121,21 +122,24 @@ void WebPermissionController::tryProcessingRequests()
 void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& origin, const WebCore::PermissionDescriptor& descriptor, WebCore::PermissionState state)
 {
+    ASSERT(isMainRunLoop());
     for (auto& observer : m_observers) {
```
`for (auto& [permissionStatusIdentifier, observer] : m_observers) {`
would look better IMO.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -121,21 +122,24 @@ void WebPermissionController::tryProcessingRequests()
-        if (observer.origin() == origin && observer.descriptor() == descriptor)
-            observer.stateChanged(state);
+        if (observer.value.origin == origin && observer.value.descriptor == descriptor)
+            WebCore::PermissionStatus::stateChanged(observer.key, observer.value.contextIdentifier, state);
```
`observer.key` -> `permissionStatusIdentifier`
`observer.value.contextIdentifier` -> `observer.contextIdentifier`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -25,17 +25,18 @@
 #pragma once
 
-#include "PermissionState.h"
+#include "PermissionDescriptor.h"
```
Can't this be forward declared?

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
         return;
     }
 
     auto* document = dynamicDowncast<Document>(context);
```
`dynamicDowncast<Document>(*context);` would avoid an unnecessary null check

RupinMittal replied:
Good point--Fixed!

RupinMittal replied:
Is `if (!context || ... )` on line 86 the unnecessary null check that we are trying to avoid?

cdumez replied:
No, otherwise I would have commented on that line. I was trying to avoid the null check done internally by `dynamicDowncast<Document>(context)` which is why I asked to change to `dynamicDowncast<Document>(*context)`. But then you made the change I suggested + dropped the earlier null checking code (which I didn't ask to do)

RupinMittal replied:
Ah I see. Sorry about that--I misunderstood the issue. I'll make the fix and push again when the changes related to the observer are done as well.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(context)) {
```
`*context` to avoid a null check.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        context->postTask([parameterDescriptor, promise = WTFMove(promise),  permissionState = WTFMove(permissionState)](auto& context) mutable {
-            if (!permissionState) {
-                promise.reject(Exception { NotSupportedError });
+            auto* document = downcast<Document>(&context);
```
`auto& document = downcast<Document>(context);`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+        return;
+    }
+
+    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
```
Why are we capturing `this` and `protectedThis`, seems we could just capture `document = Ref { document }`.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+        workerGlobalScope->thread().workerLoaderProxy().postTaskToLoader(WTFMove(completionHandler));
+        return;
+    }
+
```
Please `ASSERT(document);` here and use `document` instead of `context` below.

RupinMittal replied:
You're right--we could just ref Document, and capture that. But then would we remove the                        `!context->globalObject()` check below? If it can be removed, then maybe it can be removed from the start of the function as well? (on line 86, where we are going to remove the unneeded null check). Also, can document be used instead of context to create the `PermissionStatus` on line 159?

cdumez replied:
You can check `!document->globalObject()`, a document is a context. You can safely use document whenever context was used. Document is a subclass of ScriptExecutionContext.

RupinMittal replied:
Ah got it--and we can check `!document->globalObject()` even before the lambda right? It would be clean design to do the check in the part of the code where we have other checks relating to the document (line 92 `if (document) ...`).

cdumez replied:
Before the lambda is probably not a good idea since the lambda runs asynchronously. It is best to check the globalObject right before resolving the promise.

RupinMittal replied:
Ok sounds good. Do we prefer rejecting the promise if the globalObject check fails or it is ok to use an assert for the globalObject?. Because we expect the context to be a globalObject right? 

Also, capturing a ref of the document and using that would be the same as just capturing a ref of the context right? That would be clear and consistent with what we do the lambda in the worker case while removing the need to capture `this` and also removing the null check since we increase the ref count. We could still add the `ASSERT(document)` to ensure that this is the window case. Would we prefer that?

RupinMittal replied:
I pushed with the changes as I described above just so the bots can check the code, but I can change it based on what we decide :)

RupinMittal replied:
Turns out the `globalObject` test is only needed for getting `vm`, so the check does not need to happen anywhere else. Fixed!

cdumez commented on Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp:
```diff
@@ -0,0 +1,69 @@
+    return *m_permissions;
+}
+
+WorkerNavigatorPermissions* WorkerNavigatorPermissions::from(WorkerNavigator& navigator)
```
Why is this returning a raw pointer instead of a `WorkerNavigatorPermissions&` ? It doesn't look like it can ever return null.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -25,15 +25,17 @@
 #pragma once
 
-#include "PermissionState.h"
+// #include "PermissionDescriptor.h"
```
Should be dropped

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -47,19 +58,48 @@ Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context,
+        allPermissionStatusObjects().remove(m_identifier);
+    }
+
+    ensureOnMainThread([statusIdentifier = m_identifier]() { 
```
nit: You don't need `()`.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -47,19 +58,48 @@ Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context,
+{
+    ASSERT(isMainThread());
+
+    ScriptExecutionContext::ensureOnContextThread(contextIdentifier, [statusIdentifier, newState](auto&) mutable {
```
Why does this need to be mutable?

RupinMittal replied:
It doesn't--fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,71 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        });
+
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document.page(), [context = Ref { context }, contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                ASSERT(context.get().globalObject());
```
Why? How is this a guarantee? Note that you are making changes I didn't request between iterations and it is making it hard to review.

RupinMittal replied:
You're right--this isn't a guarantee. Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,71 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+    }
+
+    ASSERT(document);
+    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [context = Ref { *context }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
```
`document->topOrigin().data()`

cdumez replied:
Also, please capture `document` in the lambda, not `context`, as I suggested in my earlier review.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,71 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+
+    ASSERT(document);
+    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [context = Ref { *context }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+        ASSERT(context.get().globalObject());
```
Why is this a guarantee?

RupinMittal replied:
It's not, I've reverted this change to throw an exception if the check fails. Fixed!

cdumez reviewed (changes requested):
The previous iteration was super close to ready but this one seems like a step in the wrong direction.

cdumez commented on Source/WebCore/Modules/permissions/NavigatorPermissions.cpp:
```diff
@@ -25,7 +25,6 @@
 
 #include "config.h"
 #include "NavigatorPermissions.h"
-
```
This is incorrect, please keep the blank line. This is WebKit style.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionController.cpp:
```diff
@@ -25,7 +25,7 @@
 
 #include "config.h"
 #include "PermissionController.h"
-
```
Please keep blank line

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,77 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        });
+
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document.page(), [context = Ref { context }, contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                if (!context.get().globalObject()) {
```
`if (!context->globalObject())`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,77 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+
+    ASSERT(document);
+    PermissionController::shared().query(ClientOrigin { document->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [document = Ref { *document }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+        if (!document.get().globalObject()) {
```
`if (!document->globalObject()) {`

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.h:
```diff
@@ -36,24 +36,24 @@ class JSObject;
-
-    WeakPtr<Navigator> m_navigator;
+    explicit Permissions(NavigatorBase&);
+    
```
You're adding trailing spaces here :/ Ideally, this should be a blank line, without trailing spaces.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3051,7 +3051,7 @@
 		93B0A65326CDD14100AA21E4 /* NavigatorPermissions.h in Headers */ = {isa = PBXBuildFile; fileRef = 93B0A64A26CDD13E00AA21E4 /* NavigatorPermissions.h */; };
 		93B0A65426CDD14100AA21E4 /* Permissions.h in Headers */ = {isa = PBXBuildFile; fileRef = 93B0A64B26CDD13E00AA21E4 /* Permissions.h */; };
-		93B0A65626CDD14100AA21E4 /* PermissionStatus.h in Headers */ = {isa = PBXBuildFile; fileRef = 93B0A64D26CDD13F00AA21E4 /* PermissionStatus.h */; };
+		93B0A65626CDD14100AA21E4 /* PermissionStatus.h in Headers */ = {isa = PBXBuildFile; fileRef = 93B0A64D26CDD13F00AA21E4 /* PermissionStatus.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
Does this header really need to be made private? I don't see it getting included from WebKit2.

RupinMittal replied:
It's included in `WebPermissionController.cpp` since `WebPermissionController::permissionChanged` uses `WebCore::PermissionStatus::stateChanged`

RupinMittal replied:
I reverted this change since `WebPermissionController.cpp` no longer includes that file after reverting some changes made to observers. Fixed!

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -121,21 +123,24 @@ void WebPermissionController::tryProcessingRequests()
-            observer.stateChanged(state);
+    ASSERT(isMainRunLoop());
+    for (auto& [permissionStatusIdentifier, observerData] : m_observers) {
+        if (observerData.origin == origin && observerData.descriptor == descriptor)
```
It would be nice to reverse the checks since checking descriptor equality is cheaper.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,77 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
 {
-    // FIXME: support permissions in WorkerNavigator.
+    auto* context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
+    auto* document = dynamicDowncast<Document>(*context);
```
Are we sure context is non-null? If not, please add null check and early return.

cdumez replied:
Good point, I could swear an earlier iteration has the null check... :/

RupinMittal replied:
Oh wait, maybe I have misinterpreted something. I thought we were using `dynamicDowncast<Document>(*context);` to avoid a null check--as in, if context is null, then document will be null (and `workerGlobalScope` will be null when we `dynamicDowncast` cast `*context`), and since those are being checked, we'll catch the issue if context it null. So then we don't need to check if context is null explicitly. Is that not the extra null check that we were trying to remove from the previous iteration? If not, what was the unneeded null check we were trying to remove?

cdumez replied:
`dynamicDowncast<Document>(*context);` avoids a null check when we know `context` is not null. If context can be null then doing `*context` is wrong (you may be dereferencing null).

cdumez replied:
The previous iteration had a previous null check for `context` with an early return. So I suggested using `dynamicDowncast<Document>(*context);` to avoid a second unnecessary null check of the context. However, in this iteration you dropped the context (and context->globalObject()) without explanation.

RupinMittal replied:
Ah that makes sense. My apologies, I misunderstood. Fixed!

szewai commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,77 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context)) {
```
Can we put the conditions for document together? It's a bit hard to follow as current code handles document, handles worker, and then back to handle document.
Specifically, you may move the worker handling to the end and assert workerGlobalScope non-null.

cdumez replied:
I don't really understand the comment here. Rupin is first dealing with the worker case and early returning, and then dealing with the document case.

There is a little bit of document logic inside the worker case since we need to get to the document owning the worker to get its Page. However, the worker-owning document is not the same as `document` at the top-level of this function since it would be null in the worker case.

szewai replied:
There are 3 document checks outside workerGlobalScope:
```
if (document) {
         if (!document->isFullyActive()) {
```
and
`if (document && !isAllowedByFeaturePolicy(*document, permissionDescriptor.name))`
and
`ASSERT(document);`

cdumez replied:
Oh, these checks. Very true. I agree we should move the document handling with those earlier document checks.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,66 +82,77 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document.page(), [context = Ref { context }, contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                if (!context.get().globalObject()) {
+                    promise.reject(Exception { InvalidStateError, "The context is invalid"_s });
```
Is it Okay to reject promise on the main thread?

cdumez replied:
Definitely not. And it is not OK to use context above either.

Rupin, in my earlier review comments, I asked you to put this check inside the postTaskTo() lambda I believe.

RupinMittal replied:
You're right, this is not ok. I'll remove the ref and capture of context in the above lambda, and put this check inside the postTaskTo() lambda below.

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -63,6 +68,13 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
+        WebCore::ScriptExecutionContextIdentifier contextIdentifier;
+    };
+
+    HashMap<WebCore::PermissionStatusIdentifier, ObserverData> m_observers;
```
m_observers holds WebCore::PermissionObserver because PermissionObserver is not necessarily PermissionStatus by design; it can be any object that needs permission update (for example maybe Notification.permission). This has changed the design. You can avoid this change by making PermissionStatus in worker hold a mainthread PermissionObserver and register that observer.

On the other hand, if we want to make PermissionStatus the only observer, I don't think we need ObserverData or m_observers; just store the data in PermissionStatus and iterate all PermissionsStatus when state has changed.

cdumez replied:
> On the other hand, if we want to make PermissionStatus the only observer, I don't think we need ObserverData or m_observers; just store the data in PermissionStatus and iterate all PermissionsStatus when state has changed.

I recommended against this because it would be very inefficient to dispatch to another thread for each permissionStatus object (doing isolated copy of the origin each time), just to notice that this status object doesn't care about the change notification at all. The current code is way more efficient.

With regards to using a PermissionObserver, I don't have a strong opinion. The proposed code works and is simple, but it is true that we lost some functionality. It is hard for me to evaluate if this functionality is really helpful at all since this is mostly unused code at the moment. We have all this change logic implemented (and that we need to keep maintaining) even though we don't support the change event yet. Also, the proposed design would actually support the change event, it would merely not support other theoretical use cases.

szewai replied:
> I recommended against this because it would be very inefficient to dispatch to another thread for each permissionStatus object (doing isolated copy of the origin each time)

Why would we do copy each time... the PermissionStatus can get the origin from ScriptExecutionContext... we only need to save a copy in PermissionStatus for mainthread use (for searching). This proposal is basically the same as the implementation but without adding another map in PermissionController.

My point is this implementation leaves us a weird state that PermissionObserver is not used, but we make PermissionStatus subclass it and keep a map named m_observers. Either we keep the original design that PermissionObserver is used (and we will extend its usage); or we just remove PermissionObserver. I prefer the first.

cdumez replied:
Ok, I don't mind keeping PermissionObserver and making it work as long as you help Rupin implement it :P

RupinMittal replied:
Ok, so we decided to only support the observer onchange behavior for window objects in this patch. We'll do another patch later to add support for workers. The changes made to observer earlier in the history of this patch have been reverted. Fixed!

cdumez reviewed (changes requested):
Still needs work.

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -53,13 +54,15 @@ PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionSt
-
-    PermissionController::shared().addObserver(*this);
+    
+    if (isMainThread())
```
Please add a FIXME comment.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.cpp:
```diff
@@ -53,13 +54,15 @@ PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionSt
 PermissionStatus::~PermissionStatus()
 {
-    PermissionController::shared().removeObserver(*this);
+    if (isMainThread())
```
Ditto.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
 
-    auto* origin = context->securityOrigin();
-    auto originData = origin ? origin->data() : SecurityOriginData { };
+    auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context);
```
Use downcast

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-    auto* origin = context->securityOrigin();
-    auto originData = origin ? origin->data() : SecurityOriginData { };
+    auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context);
+    ASSERT(workerGlobalScope);
```
Then you can drop this assertion since downcast<> will assert for you.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-            promise.resolve(PermissionStatus::create(context, *permissionState, parameterDescriptor));
+        auto& document = downcast<Document>(context);
+        if (!document.page()) {
+            promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
```
This is still rejecting the promise on the wrong thread.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
 
-    auto* origin = context->securityOrigin();
-    auto originData = origin ? origin->data() : SecurityOriginData { };
+    auto* workerGlobalScope = downcast<WorkerGlobalScope>(context);
```
`auto& workerGlobalScope = downcast<WorkerGlobalScope>(*context);` since it cannot be null.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
-            promise.resolve(PermissionStatus::create(context, *permissionState, parameterDescriptor));
+        }
+
+        PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document.page(), [contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {                
```
`document.topOrigin().data()`

RupinMittal replied:
Fixed!

szewai reviewed (approved):
LGTM

RupinMittal committed:
Add support to the Permissions API for dedicated workers
https://bugs.webkit.org/show_bug.cgi?id=243805

Reviewed by Chris Dumez and Sihui Liu.

Permissions::query() should be exposed for both window and workers. Currently,
it is only exposed for window. This patch is the first step in exposing it for
all of the workers.

* LayoutTests/http/tests/notifications/permission/worker-permission-query-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/worker-permission-query.html: Added.
* LayoutTests/http/tests/notifications/permission/worker-permission-query.js: Added.
(onmessage):
* LayoutTests/imported/w3c/web-platform-tests/permissions/all-permissions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/storage/permission-query.https.any.worker-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/NavigatorPermissions.cpp:
(WebCore::NavigatorPermissions::permissions):
* Source/WebCore/Modules/permissions/NavigatorPermissions.h:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionStatus.cpp:
(WebCore::PermissionStatus::PermissionStatus):
(WebCore::PermissionStatus::~PermissionStatus):
(WebCore::PermissionStatus::stateChanged):
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/PermissionStatus.idl:
* Source/WebCore/Modules/permissions/PermissionStatusIdentifier.h: Added.
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::create):
(WebCore::Permissions::Permissions):
(WebCore::Permissions::navigator):
(WebCore::Permissions::query):
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/Modules/permissions/Permissions.idl:
* Source/WebCore/Modules/permissions/WorkerNavigator+Permissions.idl: Copied from Source/WebCore/Modules/permissions/Permissions.idl.
* Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.cpp: Copied from Source/WebCore/Modules/permissions/NavigatorPermissions.cpp.
(WebCore::WorkerNavigatorPermissions::WorkerNavigatorPermissions):
(WebCore::WorkerNavigatorPermissions::permissions):
(WebCore::WorkerNavigatorPermissions::from):
(WebCore::WorkerNavigatorPermissions::supplementName):
* Source/WebCore/Modules/permissions/WorkerNavigatorPermissions.h: Copied from Source/WebCore/Modules/permissions/NavigatorPermissions.h.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::addObserver):
(WebKit::WebPermissionController::removeObserver):
(WebKit::WebPermissionController::permissionChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:

Canonical link: https://commits.webkit.org/253447@main
